### ISSUE TITLE
Panic: protocol bug, message handled out of order

### ISSUE THREAD

**Issue by nqvinh00** [redis/rueidis#690]
Hi @rueian 
I encountered an error while resharding the OCI cache with v1.0.47. I'm checking the server side, too, but I want to ask if there is any cause on the client side.

panic: protocol bug, message handled out of order

goroutine 22276 [running]:
github.com/redis/rueidis.(*pipe)._backgroundRead(0xc0001c8900)
        /gopath/pkg/mod/github.com/redis/rueidis@v1.0.47/pipe.go:506 +0x931
github.com/redis/rueidis.(*pipe)._background(0xc0001c8900)
        /gopath/pkg/mod/github.com/redis/rueidis@v1.0.47/pipe.go:326 +0x88
created by github.com/redis/rueidis.(*pipe).background.func1 in goroutine 22001
        /gopath/pkg/mod/github.com/redis/rueidis@v1.0.47/pipe.go:309 +0x56

**Comment by rueian** [comment]
Hi @nqvinh00,

This usually needs tcp dumps to debug. Are you able to reproduce the situation in a separated environment?

**Comment by nqvinh00** [comment]
Hi @rueian. Tks for the quick reply, in the past, it didn't occur on AWS Elasticache. I will let you know when I get more info.

**Comment by nqvinh00** [comment]
In the result of tcpdump, my application still received unsubscribe and unsubscribe commands. Is there any guide to debug @rueian

**Comment by rueian** [comment]
Do you mean `sunsubscribe` responses? I think it might be a bug in rueidis. It would be helpful if you have an example program that can reproduce the panic.

**Comment by nqvinh00** [comment]
<img width="1264" alt="image" src="https://github.com/user-attachments/assets/3d6bae70-4ccc-4eb6-84d5-d80f174666ad">
I have checked the tcpdump to verify if the application receives the sunsubscribe command from the Redis server while resharding.

**Comment by rueian** [comment]
> The next slice of the result channel is nil.

And then the program should go into the `if unsub {` branch to ignore those `sunsubscribe` messages.

So it is weird it panic. Is the last response you received a `sunsubscribe`?

**Comment by nqvinh00** [comment]
> Do you mean `sunsubscribe` responses? I think it might be a bug in rueidis. It would be helpful if you can provide an example program that can reproduce the panic and the TCP dumps without exposing your sensitive data.

I have sent you an email with the tcpdump result. Please take a look.

**Comment by nqvinh00** [comment]
> > The next slice of the result channel is nil.
> 
> And then the program should go into the `if unsub {` branch to ignore those `sunsubscribe` messages.
> 
> So it is weird it panic. Is the last response you received a `sunsubscribe`?

If I don't miss anything so yes, the last response application received is `sunsubscribe`

**Comment by rueian** [comment]
Yes, I also checked your tcpdump and I saw only `sunsubscribe` as well. So, I guess the application panicked before these `sunsubscribe`. Do you have an earlier dump?

**Comment by rueian** [comment]
And would you like to try `v1.0.51-debug.1`? It will print the exact msg when it goes into the panic branch.

**Comment by nqvinh00** [comment]
> Yes, I also checked your tcpdump and I saw only `sunsubscribe` as well. So, I guess the application panicked before these `sunsubscribe`. Do you have an earlier dump?



> And would you like to try `v1.0.51-debug.1`? It will print the exact msg when it goes into the panic branch.

I have sent you an earlier.
It's hard to upgrade the version now, but I will consider upgrading it later.

**Comment by rueian** [comment]
> I have sent you an earlier.

I saw it, thanks, and there were some `-MOVED` indeed. But, unfortunately, I still can't find any hint from it yet.

**Comment by nqvinh00** [comment]
> > I have sent you an earlier.
> 
> I saw it, thanks, and there were some `-MOVED` indeed. But, unfortunately, I still can't find any hint from it yet.

Tks, please let me know if you find anything

**Comment by rueian** [comment]
Hi @nqvinh00, what version of OCI cache you are using? Is it a Valkey 8?

**Comment by nqvinh00** [comment]
> Hi @nqvinh00, what version of OCI cache you are using? Is it a Valkey 8?

It's redis 7.0

**Comment by nqvinh00** [comment]
After comparing the timestamp of tcpdump and the time application panicked, the application panicked right after the first `sunsubcribe`. @rueian does this info help you?

**Comment by rueian** [comment]
Hi @nqvinh00,

Yes, and the situation I believe is the same as https://github.com/valkey-io/valkey/issues/1066. So we need a way to differentiate `sunsubcribe` responses sent by the cluster proactively and sunsubcribe` responses that respond to your `sunsubcribe` requests.

I guess we need to inject a special PING after each of your `sunsubcribe` requests in order to differentiate `sunsubcribe` responses correctly.

**Comment by rueian** [comment]
Hi @nqvinh00, the fix is ready here https://github.com/redis/rueidis/pull/691, and you can try it with `v1.0.52-alpha.1`.

**Comment by nqvinh00** [comment]
> Hi @nqvinh00, the fix is ready here #691, and you can try it with `v1.0.52-alpha.1`.

Hi @rueian, when will the official version be available?

**Comment by rueian** [comment]
Usually, we release a new version every 2~3 weeks. So I guess it will be available around the end of Dec.

**Comment by nqvinh00** [comment]
Tks for your support. I will update asap and give you a feedback later.

**Comment by nqvinh00** [comment]
Hi @rueian, I have upgraded the alpha version and re-tested it 1 time. No panic occurs but I still capture tcpdump, do you need it?

**Comment by rueian** [comment]
Thanks. But I only need the dump if you find some problems.

**Comment by nqvinh00** [comment]
> Thanks. But I only need the dump if you find some problems.

It seems normal, you can close this issue. I will let you know if there's any problem.
Tks a lot.

**Comment by dangngoctam00** [comment]
@rueian 
Hi, I checked the tcpdump based on fixed version `v1.0.52-alpha.1`. There is a pattern like this:
```
1. client send unsubscribe + ping

2. client send hget

3. server unsubscribe proactively

4. service send move + pong + hget response
```
It seems that this version fix the panic error, it discard response 3 and keep the request/response sync.
Thank you.
